File: 0402-remove-k-digits/0402-remove-k-digits.java
class Solution {
    public String removeKdigits(String num, int k) {
        Stack<Character> st=new Stack<>();
        int n=num.length();
        for(int i=0;i<n;i++){
            while(!st.isEmpty() && k>0 && (st.peek()-'0')>(num.charAt(i)-'0')){
                st.pop();
                k--;
            }
            st.push(num.charAt(i));
        }
        while(k>0){
            st.pop();
            k--;
        }
        if(st.isEmpty()) return "0";
        StringBuilder res=new StringBuilder();
        while(!st.isEmpty()){
            res.append(st.pop());
        }
        while(res.length()!=0 && res.charAt(res.length()-1)=='0') res.deleteCharAt(res.length()-1);
        res.reverse();
        if(res.isEmpty()) return "0";
        return res.toString();
    }
}

File: 0402-remove-k-digits/0402-remove-k-digits.py
class Solution:
    def removeKdigits(self, num: str, k: int) -> str:
        stack=[]
        n=len(num)
        for i in range(n):
            while stack and k>0 and (ord(stack[-1])-ord('0'))>(ord(num[i])-ord('0')):
                stack.pop()
                k-=1
            stack.append(num[i])
        while k>0:
            stack.pop()
            k-=1
        if not stack:
            return "0"
        res=[]
        while stack:
            res.append(stack.pop())
        while res and res[-1]=='0':
            res.pop()
        res.reverse()
        if not res:
            return "0"
        return ''.join(res)